ARCs/arc-draft_asa_url_templating.md

```diff
@@ -0,0 +1,197 @@
+5. _Hash type_ **MUST** be 'sha2-256' but is explicitly specified so as to allow future versions the option of using other hashes. 
+
+> IPFS may add future versions of the cid spec, and add additional multicodec types or other supported hashes.  Implementations SHOULD use IFPS libraries where possible that accept multicodec and hash types as named values and allow a CID to be composed generically.
+
+### Examples
+
+> ASA URL: template-ipfs://{ipfscid:0:dag-pb:reserve:sha2-256}/arc3.json  
+> ASA URL: template-ipfs://{ipfscid:1:raw:reserve:sha2-256}  
+> ASA URL: template-ipfs://{ipfscid:1:dag-pb:reserve:sha2-256}/metadata.json  
+
+#### Deployed Testnet Example
+
+An example was pushed to Testnet, converting from an existing ARC3 mainnet ASA:  
+https://algoexplorer.io/asset/560421434  
+With IPFS URL: ```ipfs://QmQZyq4b89RfaUw8GESPd2re4hJqB8bnm4kVHNtyQrHnnK```
+
+The Testnet ASA was minted with the URL: ```template-ipfs://{ipfscid:0:dag-pb:reserve:sha2-256}``` as this CID is a V0 / dag-pb CID.
+
+A helpful link to 'visualize' CIDs and for this specific id, is: https://cid.ipfs.io/#QmQZyq4b89RfaUw8GESPd2re4hJqB8bnm4kVHNtyQrHnnK
+
+Using the using the example encoding implementation, results in virtual 'reserve address' of **EEQYWGGBHRDAMTEVDPVOSDVX3HJQIG6K6IVNR3RXHYOHV64ZWAEISS4CTI**
+
+As can be seen here: https://testnet.algoexplorer.io/asset/66753108
+
+Using the forked https://github.com/TxnLab/arc3.xyz repo, with testnet selected, and the /nft/66753108 url - the browser will display the original content as-is, using only the Reserve address as the source of the content hash.
+
+## Rationale
+The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages.
+
+## Backwards Compatibility
+The 'template-' prefix of the scheme is intended to break clients reading these ASA URLS outright.  Clients interpreting these URLS as-is would likely yield unusual errors.  Code checking for an explicit 'ipfs' scheme for example will not see this as compatible with any of the default processing and SHOULD treat the URL as if it were simply unknown/empty.
+
+## Reference Implementation
+
+### Encoding
+#### Go[lang] implementation
+
+```go
+import (
+    "github.com/algorand/go-algorand-sdk/types"
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multihash"
+)
+// ...
+func ReserveAddressFromCID(cidToEncode cid.Cid) (string, error) {
+    decodedMultiHash, err := multihash.Decode(cidToEncode.Hash())
+    if err != nil {
+        return "", fmt.Errorf("failed to decode ipfs cid: %w", err))
+    }
+    return types.EncodeAddress(decodedMultiHash.Digest)
```

Comment:
I'm slightly concerned that the 4-byte checksum added would cause confusion for decoding but I guess it's okay if it's used in an address field like `reserve`. I wonder how the encoding would be handled for the note field though.

Reply:
> You probably have a specific use case for this proposed ARC but I think ipns also has its place for mutable NFTs. Curious why ipns is entirely left out of the conversation for NFT mutability. Haven't seen it in any ARCs so far or maybe I've missed it....

Correct.  I looked at IPNS originally and intended to use it, but talking w/ IPFS devs, it's not really meant for large-scale use and has many problems, particularly with scale.  The key is effectively fixed to a particular clusters key and if referencing, for example, millions of 'files' off that single mutable ipns root then any change to any file requires a republishing of the entire root dag node and republishing the 'single' ipns resolution link.  
So, providing distinct, independent mutable content isn't appropriate for single ipns resolution, and on top of that, resolving ipns is extremely slow (as they admit).

Reply:
> I'm slightly concerned that the 4-byte checksum added would cause confusion for decoding but I guess it's okay if it's used in an address field like `reserve`. I wonder how the encoding would be handled for the note field though.

Oh, are you referring to the Algorand's 'display' of an Algorand Address based off the raw 32-byte public key where it adds a checksum and uses that for the 58 byte base32 'public account address?'  
Correct, it's completely ignored (and is within algorand as well - other than as input 'validation').  The SDK will handle going back and forth and as long as developers use the base 32-bytes of the PK it should work fine.

Reply:
..continuing w/ IPNS point (going off the confused emoji response).
If you had an ASA URL of ipns://xxxxxxxx/somename.json yet you had millions of 'somename' (I will in my case at least) then you basically have a single IPFS DAG at root (imagine a single directory with a million files) with nodes for every file off it.  To add/change any file within that virtual directory - the hash of the node changing changes, as well as the creation of new root node (referencing ALL the children) and its new hash.  That single new 'root' hash has to then be published as "the" new IPNS hash to represent that ipns:// >- xxxx <-- link.  
It's 'ok' if you're doing for single root level changes but since the base ipns key that you're providing a replacement for is explicitly part of the configuration of the ipfs node itself, you really can't manage millions of those base entries like you'd want per ASA.

Reply:
I am annoyed to hear that ipns doesn't scale and is slow. 😤 Your proposal looks like a decent solution to this problem and I might use it at some point as ipns wouldn't suit my use case with what you have described. For now, without testing the implementation, I don't have any further comments.